Fix versions.props handling of partially overlapping forces

dansanduleac opened the pull request:
Fixes #379 properly

## Before this PR

In `checkBomConflict`, multiple forces from `versions.props` that match the same artifact would pick the longest force as the one that was "used".
This is not always the the one that nebula picks, because nebula just picks the last line that matches the artifact.

A similar mistake occurs in `checkNoUnusedPin` where _every_ force is individually treated as used if it matches even a single artifact, disregarding other forces in the file that might have overridden it.

## After this PR

* `checkBomConflict`: Resolve conflicts by associating an artifact with the last `versions.props` line that matches it.
* `checkNoUnusedPin`: when considering a resolved artifact, only treat the *last* matching force as used.

This now matches nebula's behaviour.

dansanduleac committed:
Fix versions.props line disambiguation

ferozco commented:
discussed offline, would like to see an additional test showing the previously broken behaviour

dansanduleac committed:
Fix which force is used logic in CheckNoUnusedPin too

dansanduleac committed:
Fix which force is used logic in CheckNoUnusedPin too

dansanduleac committed:
fix tests

dansanduleac committed:
fix tests

dansanduleac committed:
better test

dansanduleac committed:
Merge branch 'develop' into ds/fix-resolve-conflicts-bom

dansanduleac commented:
@ferozco actually quite hard to make up a test for `checkBomConflict` that doesn't also end up with unused forces that get picked up by the other task.
In the meanwhile though, I noticed `checkNoUnusedPins` was also broken in a similar way.
Updated the description, fixed it and added tests for that ([here](https://github.com/palantir/gradle-baseline/pull/423/files#diff-1b3ac1518abcb74fddedc83185b367e8R196)).

As part of this, noticed the original test 'Version props conflict should succeed' was wrong! The first star force wasn't actually being used anywhere, so the new logic correctly marked it as unused. I therefore changed the test to use jackson and make sure the star force is also used.

dansanduleac commented on gradle-baseline-java/src/test/groovy/com/palantir/baseline/BaselineVersionsIntegrationTest.groovy:
```diff
@@ -97,9 +97,9 @@ class BaselineVersionsIntegrationTest  extends AbstractPluginTest {
+    void buildAndFailWith(String error) {
         BuildResult result = with('checkVersionsProps', '--stacktrace').buildAndFail()
-        result.output.contains(error)
+        assert result.output.contains(error)
```
better error messages! groovy will actually destructure your command when you use assert, just like it does when putting boolean expressions in a `then:` / `expect:` block.

dansanduleac commented on gradle-baseline-java/src/test/groovy/com/palantir/baseline/BaselineVersionsIntegrationTest.groovy:
```diff
@@ -174,21 +175,42 @@ class BaselineVersionsIntegrationTest  extends AbstractPluginTest {
+        buildFile << standardBuildFile(projectDir)
+        buildFile << """
         dependencies {
-            compile 'com.google.guava:guava'
```
original test was failing with the newly corrected logic, because `com.google.guava:*` is not being used by any artifact.

dansanduleac commented:
@ferozco had an even better idea of what we can do here.
To avoid PR inflation, I'll file an issue with the follow-up.
